src/vsr/journal.zig

```diff
@@ -1212,13 +1177,46 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
         /// 4. has command=reserved or command=prepare
         fn recover_slots(journal: *Journal) void {
             const replica = @fieldParentPtr(Replica, "journal", journal);
+            const log_view = replica.superblock.working.vsr_state.log_view;
+            const view_change_headers = replica.superblock.working.vsr_headers();
 
             assert(journal.status == .recovering);
             assert(journal.reads.executing() == 0);
             assert(journal.writes.executing() == 0);
             assert(journal.dirty.count == slot_count);
             assert(journal.faulty.count == slot_count);
 
+            // Discard headers which we are certain do not belong in the current log_view.
+            // - This ensures that we don't accidentally set our new head op to be a message
+            //   which was truncated but not yet overwritten.
+            // - This is also necessary to ensure that generated DVC's headers are complete.
+            //
+            // It is essential that this is performed before we compute the op_max so that the
+            // recovery cases apply correctly.
+            for (journal.headers_redundant) |*header_untrusted, index| {
+                const slot = Slot{ .index = index };
+                if (header_ok(replica.cluster, slot, header_untrusted)) |header| {
+                    var view_range = view_change_headers.view_for_op(header.op, log_view);
+                    view_range.max = std.math.min(view_range.max, log_view);
+
+                    if (header.command == .prepare and !view_range.contains(header.view)) {
+                        header_untrusted.* = Header.reserved(replica.cluster, index);
+                    }
+                }
+            }
+
+            for (journal.headers) |*header_untrusted, index| {
```

Comment:
Can we combine these loops?

```
for (.{journal.headers_redundant, journal.headers}) |headers| {
   ...
}
```

Reply:
Yes, though it ends up being
```
            for ([_][]align(constants.sector_size) Header{
                journal.headers_redundant,
                journal.headers,
            }) |headers| {
```
to satisfy a zig type inference bug.